File: amp-sdk/go/client/eip712.go
package client

import (
	"fmt"
	"math/big"

	"golang.org/x/crypto/sha3"
)

// ComputeOutcomeEip712Digest computes the canonical EIP-712 digest over
// (matchId, outcome, transcriptHash) that the AMP server verifies in
// submitOutcome and the verifier signs.
//
// Byte-identical to AMP-Server::compute_outcome_eip712_digest and the
// Rust/C#/Python/JS helpers. The shared known-answer vector is enforced in
// the test below and across the other SDKs.
//
//   - matchId: if it parses as a decimal uint256 it is used directly; otherwise
//     it is keccak256(utf8).
//   - outcome: the victor index (1..=4).
//   - transcriptHash: the 32-byte keccak256 transcript/replay hash.
//   - chainID: the EIP-712 domain chain id (e.g. 43113 for Fuji).
//   - verifyingContract: the 20-byte AMPSettlement contract address.
func ComputeOutcomeEip712Digest(matchID string, outcome uint8, transcriptHash []byte, chainID uint64, verifyingContract []byte) ([32]byte, error) {
	var out [32]byte
	if len(transcriptHash) != 32 {
		return out, fmt.Errorf("transcript_hash must be exactly 32 bytes, got %d", len(transcriptHash))
	}
	if len(verifyingContract) != 20 {
		return out, fmt.Errorf("verifying_contract must be exactly 20 bytes, got %d", len(verifyingContract))
	}

	asyncResultTypehash := keccak256([]byte("AsyncResult(uint256 matchId,uint8 outcome,bytes32 transcriptHash)"))

	// matchId → uint256 (decimal parse) else keccak256(utf8).
	var matchIDWord [32]byte
	if v, ok := new(big.Int).SetString(matchID, 10); ok {
		v.FillBytes(matchIDWord[:])
	} else {
		matchIDWord = keccak256([]byte(matchID))
	}

	var outcomeWord [32]byte
	outcomeWord[31] = outcome

	// structHash = keccak256(abi.encode(typeHash, matchId, outcome, transcriptHash))
	structInput := make([]byte, 0, 128)
	structInput = append(structInput, asyncResultTypehash[:]...)
	structInput = append(structInput, matchIDWord[:]...)
	structInput = append(structInput, outcomeWord[:]...)
	structInput = append(structInput, transcriptHash...)
	structHash := keccak256(structInput)

	domainTypehash := keccak256([]byte("EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"))
	nameHash := keccak256([]byte("AMPSettlement"))
	versionHash := keccak256([]byte("1"))

	var chainWord [32]byte
	new(big.Int).SetUint64(chainID).FillBytes(chainWord[:])

	var addrWord [32]byte
	copy(addrWord[12:], verifyingContract)

	domainInput := make([]byte, 0, 160)
	domainInput = append(domainInput, domainTypehash[:]...)
	domainInput = append(domainInput, nameHash[:]...)
	domainInput = append(domainInput, versionHash[:]...)
	domainInput = append(domainInput, chainWord[:]...)
	domainInput = append(domainInput, addrWord[:]...)
	domainSeparator := keccak256(domainInput)

	digestInput := append([]byte{0x19, 0x01}, domainSeparator[:]...)
	digestInput = append(digestInput, structHash[:]...)
	return keccak256(digestInput), nil
}

func keccak256(b []byte) [32]byte {
	h := sha3.NewLegacyKeccak256()
	h.Write(b)
	var out [32]byte
	h.Sum(out[:0])
	return out
}

File: amp-sdk/js/src/eip712.ts
/**
 * EIP-712 typed-data digest for AMP match outcomes.
 *
 * Mirrors `compute_outcome_eip712_digest` in amp-server/src/main.rs and the
 * helpers in amp-sdk/{csharp,python}/. The digest MUST be byte-identical
 * across all languages; the cross-language KAT
 * (`test_outcome_digest_known_vector_cross_lang` in amp-server) fails if any
 * SDK diverges.
 *
 * Known-answer test vector (chain_id=43113, verifying_contract=0x000...000):
 *   matchId="1", outcome=1, transcript_hash=zero[32] =>
 *   2d2525ad5098ca8f82a2a6cabc6775c40a55df96dfa2fbb46d7c0e372b99096c
 */

import { TypedDataEncoder, keccak256, toUtf8Bytes, getBytes } from "ethers";

/** Default chain ID for the AMP settlement contract (Fuji testnet). */
export const DEFAULT_CHAIN_ID = 43113n;

/** Default verifying contract address (zero — override at app startup). */
export const DEFAULT_VERIFYING_CONTRACT = "0x0000000000000000000000000000000000000000";

/** Canonical domain name & version, fixed by the on-chain verifier. */
export const DOMAIN_NAME = "AMPSettlement";
export const DOMAIN_VERSION = "1";

/** Default EIP-712 domain parameters. Override at app startup. */
export interface Eip712DomainParams {
  chainId: bigint;
  verifyingContract: string;
}

/**
 * Compute the canonical EIP-712 digest over (matchId, outcome, transcriptHash).
 * Sign this digest with the player's wallet; the server will recover the
 * address and require it to match the authenticated player.
 *
 * @param matchId Match identifier as a UTF-8 string (numeric or UUID-like).
 *                Numeric strings are encoded as uint256 directly; others are
 *                keccak256-hashed to fit uint256.
 * @param outcome Victor index, 1..=4 (server invariant).
 * @param transcriptHash 32-byte keccak256 of the match transcript.
 * @param domain Override chain ID / verifying contract to match deployment.
 */
export function computeOutcomeEip712Digest(
  matchId: string,
  outcome: number,
  transcriptHash: Uint8Array,
  domain: Eip712DomainParams = {
    chainId: DEFAULT_CHAIN_ID,
    verifyingContract: DEFAULT_VERIFYING_CONTRACT,
  },
): Uint8Array {
  if (transcriptHash.length !== 32) {
    throw new Error(`transcript_hash must be exactly 32 bytes, got ${transcriptHash.length}`);
  }
  if (outcome < 1 || outcome > 4) {
    throw new Error(`outcome must be 1..=4 (server invariant), got ${outcome}`);
  }

  // matchId → uint256 (parse decimal, else keccak256-hash UTF-8 bytes).
  const matchIdU256 = tryParseUint256(matchId) ?? BigInt(keccak256(toUtf8Bytes(matchId)));

  const types = {
    AsyncResult: [
      { name: "matchId", type: "uint256" },
      { name: "outcome", type: "uint8" },
      { name: "transcriptHash", type: "bytes32" },
    ],
  };
  const value = {
    matchId: matchIdU256,
    outcome,
    transcriptHash: transcriptHash,
  };
  const domainObj = {
    name: DOMAIN_NAME,
    version: DOMAIN_VERSION,
    chainId: domain.chainId,
    verifyingContract: domain.verifyingContract,
  };
  // TypedDataEncoder.hash returns the canonical EIP-712 digest
  // (0x1901 || domainSeparator || structHash, keccak256'd).
  return getBytes(TypedDataEncoder.hash(domainObj, types, value));
}

function tryParseUint256(s: string): bigint | null {
  try {
    const v = BigInt(s);
    if (v < 0n || v > 2n ** 256n - 1n) return null;
    return v;
  } catch {
    return null;
  }
}
